### ISSUE TITLE
Some Clients Can't Connect to Servers

### ISSUE THREAD

**Issue by brucethemoose** [PrinceOfAmber/Cyclic#1437]
Minecraft Version: 1.15.2

Forge Version: 31.2.0

Mod Version:  Cyclic-1.15.2-0.4.1.jar  1.19.14

Single Player or Server: Server

Describe problem (what you were doing; what happened; what should have happened):

Cyclic prevents some clients from connecting to servers.  Some backstory: I, and a few other players, were consistently failing to connect to modded 1.15.2 servers running different modpacks, while other players can seemingly connect just fine. When connecting to a public server, the client hangs at "Encrypting" for awhile (~30 seconds?), and then simply receives a "disconnected" message:

![Capture](https://user-images.githubusercontent.com/8422224/83557146-242fdc00-a4df-11ea-8930-546b4e697236.JPG)

I wrote this off as a local issue at first, but it happened to me on both Windows and Linux, across multiple servers, with others reporting similar problems. Then someone mentioned Cyclic as a culprit, so I started up 2 local 1.15.2 instances on the same machine, side by side. With no internet, and nothing but Forge, the instances connect to each other just fine:

![JustForge](https://user-images.githubusercontent.com/8422224/83557420-9bfe0680-a4df-11ea-8976-879a97c0d878.jpg)

But with Cyclic and Forge, the client hangs, just like it does with a remote server, albeit with a different error message:
![JustCyclic1](https://user-images.githubusercontent.com/8422224/83557466-b6d07b00-a4df-11ea-9e26-54776ae89e28.JPG)
![Cyclic2](https://user-images.githubusercontent.com/8422224/83557492-c354d380-a4df-11ea-983c-fd761dbd2d77.JPG)

This happens on Both Windows 10 and Manjaro Linux 20.1, with MultiMC and Twitch as launchers, but I don't have another computer to test it on. 

Log file link: 

Client: https://paste.ee/p/oQyDd

Server: https://paste.ee/p/hPAWJ

**Comment by Mitsuma** [comment]
Just commenting to confirm this bug, happens in most 1.15 packs such as ATM5.  
Removal of Cyclic lets me join the server again.

**Comment by Darkere** [comment]
Some quick profiling shows this is due to the baking of the item and fluid pipes. Unfortunately that's as far as I go.. Cause I've got 0 clue how that works. 

Profiled just joining a server with just cyclic installed. 

![image](https://user-images.githubusercontent.com/4283717/83573275-6979fb00-a52b-11ea-8585-a3849d2fbc09.png)


[profilingcyclicjoinserver.zip](https://github.com/Lothrazar/Cyclic/files/4719848/profilingcyclicjoinserver.zip)

**Comment by JulianGmp** [comment]
fyi, over at [All the Mods 5](https://github.com/AllTheMods/ATM-5) this issue was observed too. 
See: 
* https://github.com/AllTheMods/ATM-5/issues/144#issuecomment-639158398
* https://github.com/AllTheMods/ATM-5/issues/143#issuecomment-639138232

**Comment by Lothrazar** [comment]
Wow lots of details on the issue this is great.

One of the logs simply has an overloaded message followed timeout message

 Can't keep up! Is the server overloaded? Running 5773ms or 115 ticks behind
name=brucethemoose2,properties={},legacy=false] (/..) lost connection: Timed out

Which isnt too helpful, but ill check out the profile

EDIT: Just did a quick smoke test on forge 31.2.5 . started and joined a (local LAN) server, joined ok placed down some fluid pipes and things, left and joined again. cant get it to break yet

**Comment by eliminat** [comment]
My multiplayer world is showing nothing cyclic in the logs, we have no cyclic pipes or wires. Only a couple cyclic block placers, yet routinely have trouble connecting with it installed.
So doesn't appear limited to a pipe/wire thing. However, it's also not 100% for us.
Difficult to confirm, but it *seems* ( yes, I hate subjectivity in problem reports too) the longer the client is running, the more likely it is to run into the issue. If I hit connect as soon as possible, I see more likely to get a successful connection. If I step away while it's loading and it sits for a minute or 2 I may not be able to connect until I restart the client. Yet I also have a friend who seems to never have the issue.

But Cyclic removed, no more issue connecting for anybody.

```
Sample from server while connection is attempted:
[12:43:45] [Netty Epoll Server IO #11/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[12:43:46] [User Authenticator #23/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[12:44:18] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@2220ea50[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@2db88e71]},legacy=false] (/192.168.1.100:58782) lost connection: Timed out
[12:44:29] [Netty Epoll Server IO #12/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[12:44:30] [User Authenticator #24/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[12:45:02] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@4b28046c[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@74cfa58a]},legacy=false] (/192.168.1.100:59890) lost connection: Timed out
[12:45:15] [Netty Epoll Server IO #13/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[12:45:16] [User Authenticator #25/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[12:45:48] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@32090d15[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@50684bf7]},legacy=false] (/192.168.1.100:60984) lost connection: Timed out
[12:46:10] [Netty Epoll Server IO #14/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[12:46:10] [User Authenticator #26/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[12:46:42] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@4c6039d5[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@4475bf1a]},legacy=false] (/192.168.1.100:34078) lost connection: Timed out
[12:48:26] [Netty Epoll Server IO #15/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[12:48:27] [User Authenticator #27/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[12:48:59] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@30d1851a[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@17bc2515]},legacy=false] (/192.168.1.100:37350) lost connection: Timed out
[13:00:00] [Server thread/INFO] [tombstone/]: The player  has successfully been saved
[13:04:30] [Netty Epoll Server IO #0/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[13:04:30] [User Authenticator #28/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[13:05:02] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@275913e1[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@44f5398c]},legacy=false] (/192.168.1.100:60716) lost connection: Timed out
[13:05:12] [Netty Epoll Server IO #1/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[13:05:13] [User Authenticator #29/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[13:05:45] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@29f56687[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@65964e11]},legacy=false] (/192.168.1.100:33506) lost connection: Timed out
[13:05:56] [Netty Epoll Server IO #2/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[13:05:56] [User Authenticator #30/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[13:06:28] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@64533644[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@4843bf9a]},legacy=false] (/192.168.1.100:34546) lost connection: Timed out
[13:06:47] [Netty Epoll Server IO #3/INFO] [co.li.co.ConfigSyncManager/]: Skipping config sync, No mods have registered a syncable config.
[13:06:48] [User Authenticator #31/INFO] [minecraft/ServerLoginNetHandler]: UUID of player eliminat is xxxxxxxx-eef0-4223-a221-78fbe5ff3406
[13:07:20] [Server thread/INFO] [minecraft/ServerLoginNetHandler]: com.mojang.authlib.GameProfile@1de4786c[id=xxxxxxxx-eef0-4223-a221-78fbe5ff3406,name=eliminat,properties={textures=[com.mojang.authlib.properties.Property@52111acb]},legacy=false] (/192.168.1.100:35778) lost connection: Timed out
```

**Comment by Darkere** [comment]
@Lothrazar My profiling was a fresh server without anything placed. 

For any local servers this will be fine. Likely small packs will also be fine. The issue is the connection timing out because the client is busy. It will just take a significantly longer time to join. 

My Ryzen 3600 took 5 seconds, I can only imagine how long this takes for slower CPU's. I believe this code is just running at the wrong time or is highly inefficient.

**Comment by Goatle** [comment]
I was also running into this issue when attempting to throw together a modpack. Removing Cyclic seems to have stopped the issue from occurring as others said. I never saw anything in the log, in either the server or client side, just that the client would disconnect upon encrypting.

Seems to happen (anecdotal) when you have a large number of mods that need to send "parts" to the client that increases the base load time.

E.X. (Clientside) : [14:46:19] [Netty Client IO #0/INFO]: Read 127 parts from server

I would always notice in my client/server logs the disconnect always happened during the injection of those parts.

Forgive my terminology I do not develop mods.

**Comment by brucethemoose** [comment]
> @Lothrazar My profiling was a fresh server without anything placed.
> 
> For any local servers this will be fine. Likely small packs will also be fine. The issue is the connection timing out because the client is busy. It will just take a significantly longer time to join.

I ran into the issue running with 2 minecraft clients on the same computer, with no other mods, so its not necessarily a remote or local server issue either.

I should add that I was running an A8-3500M at 2GHz, which is a very slow CPU. But its still a true quad core, and it wasnt burdened by an old Intel IGP or insufficient RAM for the instances, so I dont think its an unrealistic scenario.

**Comment by brucethemoose** [comment]
Possible related issue in another mod: https://github.com/TheRandomLabs/RandomPatches/issues/94

EDIT: In hindsight, probably not related.

**Comment by brucethemoose** [comment]
I went back to some older Cyclic releases, and found the breaking version. Two Clients on Cyclic 0.3.0 can connect, clients on Cyclic 0.3.1 cannot. 

Logs from 0.3.1:
https://paste.ee/p/aqrGW
https://paste.ee/p/KZzRB

I've never compiled an MC mod before, but I could give a stab at building some commits between April 3 and April 7... otherwise, feel free to ping me/send me JARs @Lothrazar , as I can consistently reproduce the issue :)

**Comment by Darkere** [comment]
@Lothrazar looks like this issue can be fixed by caching the voxel shapes. See https://github.com/refinedmods/refinedstorage/commit/a9bfe70587fdea0b5c5c253ede4ae3908793a8b6 for how RS fixed the issue.

**Comment by Darkfiend009** [comment]
To add more to this, i've added the new Cyclic port to my 1.16.1 modpack. Some players are now getting the "Encrypting" message and then being disconnected, essentially timing out as the OP had described.

**Comment by Lothrazar** [comment]
Looking into the VoxelShapes solution. So far all of them are final statics except in BlockCable, i think.  Still sucks that forge has no logging about the error

**Comment by Darkere** [comment]
Looking at the profiling I did it does seem like it is caused by the things in cyclic.block.cable

So I assume those are the ones causing the problem. I can do a test with any new builds if you need it.

**Comment by Lothrazar** [comment]
Fixed in 1.16, luckily RS is using MIT license so i can basically just rip their cache strategy.  https://github.com/refinedmods/refinedstorage/blob/mc1.16/LICENSE.md  

Ill release 1.16 tonight and maybe 1.15 tomorrow
